From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etil <81570777+etil2jz@users.noreply.github.com>
Date: Sun, 23 Jan 2022 17:37:55 +0100
Subject: [PATCH] Mirai Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 5d1668c6f368843ec071df7d65dbd0a29be545d0..9f878a65f7a9da2730514a4a95d79710015ba307 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1909,6 +1909,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Mirai start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getMiraiConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Mirai end
+
         /**
          * Sends the component to the player
          *
