From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etil <81570777+etil2jz@users.noreply.github.com>
Date: Sat, 23 Oct 2021 16:12:18 +0200
Subject: [PATCH] Mirai Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index c34cfba8f9ed7e9dbd0b6b8ffef9fba46abff046..bcd74d183b60890c325d9c20e563423bc24cc791 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1714,6 +1714,18 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
             throw new UnsupportedOperationException("Not supported yet.");
         }
         // Paper end
+		
+		// Mirai start
+		@NotNull
+		public org.bukkit.configuration.file.YamlConfiguration getMiraiConfig() {
+			throw new UnsupportedOperationException("Not supported yet.");
+		}
+		
+		@NotNull
+		public java.util.Properties getServerProperties() {
+			throw new UnsupportedOperationException("Not supported yet.");
+		}
+		// Mirai end
 
         /**
          * Sends the component to the player
