From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etil <81570777+etil2jz@users.noreply.github.com>
Date: Sun, 19 Sep 2021 18:00:34 +0200
Subject: [PATCH] Mirai Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1db87711dd9b67bcba42e731c8c1a81502abf56c..b6f0d1e67bb4599a603ff7ea7adffdba07759c3f 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1714,6 +1714,18 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
             throw new UnsupportedOperationException("Not supported yet.");
         }
         // Paper end
+		
+		// Mirai start
+		@NotNull
+		public org.bukkit.configuration.file.YamlConfiguration getMiraiConfig() {
+			throw new UnsupportedOperationException("Not supported yet.");
+		}
+
+             @NotNull
+             public java.util.Properties getServerProperties() {
+               throw new UnsupportedOperationException("Not supported yet.");
+             }
+		// Mirai end
 
         /**
          * Sends the component to the player
