From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etil <81570777+etil2jz@users.noreply.github.com>
Date: Fri, 12 Nov 2021 17:21:07 +0100
Subject: [PATCH] Patina Patches


diff --git a/build.gradle.kts b/build.gradle.kts
index 92b612126a6bac0b89198a92bbb73b742ec9d064..fdc94c58570d7eaf430e9c6023abf91a4df61397 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -91,6 +91,8 @@ tasks.jar {
 }
 
 tasks.withType<Javadoc> {
+	(options as StandardJavadocDocletOptions).encoding = "UTF-8"
+	(options as StandardJavadocDocletOptions).addBooleanOption("html5", true)
     inputs.files(apiAndDocs)
         .ignoreEmptyDirectories()
         .withPropertyName(apiAndDocs.name + "-configuration")
diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index 57a21495843f3a144cd73473cdc8781d6129b7ca..321c3ce52b60004304bb48e815501cf762c1bd0c 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -1,6 +1,6 @@
 package org.bukkit.command.defaults;
 
-import com.destroystokyo.paper.util.VersionFetcher; // Paper - version supplier
+//import com.destroystokyo.paper.util.VersionFetcher; // Paper - version supplier // Patina - remove
 import com.google.common.base.Charsets;
 import com.google.common.collect.ImmutableList;
 import com.google.common.io.Resources;
@@ -28,7 +28,8 @@ import org.bukkit.util.StringUtil;
 import org.jetbrains.annotations.NotNull;
 
 public class VersionCommand extends BukkitCommand {
-    private VersionFetcher versionFetcher;
+    /*
+	private VersionFetcher versionFetcher;
     private VersionFetcher getVersionFetcher() { // lazy load because unsafe isn't available at command registration
         if (versionFetcher == null) {
             versionFetcher = Bukkit.getUnsafe().getVersionFetcher();
@@ -36,6 +37,7 @@ public class VersionCommand extends BukkitCommand {
 
         return versionFetcher;
     }
+	*/ // Patina - remove
 
     public VersionCommand(@NotNull String name) {
         super(name);
@@ -52,7 +54,8 @@ public class VersionCommand extends BukkitCommand {
 
         if (args.length == 0) {
             //sender.sendMessage("This server is running " + Bukkit.getName() + " version " + Bukkit.getVersion() + " (Implementing API version " + Bukkit.getBukkitVersion() + ")"); // Paper - moved to setVersionMessage
-            sendVersion(sender);
+            //sendVersion(sender);
+			sender.sendMessage("This server is running " + Bukkit.getName() + " version " + Bukkit.getBukkitVersion() + " (" + Bukkit.getVersion() + ")");
         } else {
             StringBuilder name = new StringBuilder();
 
@@ -155,7 +158,8 @@ public class VersionCommand extends BukkitCommand {
         return ImmutableList.of();
     }
 
-    private final ReentrantLock versionLock = new ReentrantLock();
+    /*
+	private final ReentrantLock versionLock = new ReentrantLock();
     private boolean hasVersion = false;
     private Component versionMessage = null; // Paper
     private final Set<CommandSender> versionWaiters = new HashSet<CommandSender>();
@@ -235,6 +239,7 @@ public class VersionCommand extends BukkitCommand {
         }
          */
         // Paper end
+	/*
     }
 
     // Paper start
@@ -284,4 +289,5 @@ public class VersionCommand extends BukkitCommand {
             return -1;
         }
     }
+	*/
 }
