From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etil <81570777+etil2jz@users.noreply.github.com>
Date: Fri, 12 Nov 2021 20:57:20 +0100
Subject: [PATCH] Bump Netty version to 4.1.70.Final


diff --git a/build.gradle.kts b/build.gradle.kts
index c76ddc77c09e39d2476a91de39ca55a1e737379c..ba1c53958a3892bbd73b68c98820be8e3ef6dae3 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -53,7 +53,7 @@ dependencies {
     runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.0")
     runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.0")
 
-    implementation("io.netty:netty-all:4.1.65.Final") // Paper
+    implementation("io.netty:netty-all:4.1.70.Final") // Paper // Mirai
 
     implementation("net.fabricmc:mapping-io:0.3.0") // Paper - needed to read mappings for stacktrace deobfuscation
     implementation("com.velocitypowered:velocity-native:1.1.0-SNAPSHOT") // Paper
