From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etil <81570777+etil2jz@users.noreply.github.com>
Date: Tue, 26 Oct 2021 20:17:49 +0200
Subject: [PATCH] Bump Netty version to 4.1.69.Final


diff --git a/build.gradle.kts b/build.gradle.kts
index e0077c83202e9b092bdd76c9dfbeb6f0b1c85947..054ed3a83a5d3bb392571f538f148b83fb2aa94e 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -57,7 +57,7 @@ dependencies {
     runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.0")
 
     implementation("co.aikar:cleaner:1.0-SNAPSHOT") // Paper
-    implementation("io.netty:netty-all:4.1.65.Final") // Paper
+    implementation("io.netty:netty-all:4.1.69.Final") // Paper
 
     implementation("org.quiltmc:tiny-mappings-parser:0.3.0") // Paper - needed to read mappings for stacktrace deobfuscation
     implementation("com.velocitypowered:velocity-native:1.1.0-SNAPSHOT") // Paper
