From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etil <81570777+etil2jz@users.noreply.github.com>
Date: Mon, 20 Sep 2021 21:22:25 +0200
Subject: [PATCH] (Purpur) Lagging threshold


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 5e45bc56256b0ad30e5847ceaac2769ef8bddbc2..94aa1760d0a685c39a85c0e6a57b0218ea6a1494 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -306,6 +306,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<Runnab
     public static final int TICK_TIME = 1000000000 / MinecraftServer.TPS;
     private static final int SAMPLE_INTERVAL = 20; // Paper
     public final double[] recentTps = new double[ 3 ];
+	public boolean lagging = false; // Purpur
     public final SlackActivityAccountant slackActivityAccountant = new SlackActivityAccountant();
     // Spigot end
     public static long currentTickLong = 0L; // Paper
@@ -1265,6 +1266,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<Runnab
                         this.recentTps[1] = tps5.getAverage();
                         this.recentTps[2] = tps15.getAverage();
                         // Paper end
+						lagging = recentTps[0] < xyz.arthurb.mirai.MiraiConfig.laggingThreshold; // Purpur
                         tickSection = curTime;
                     }
                     // Spigot end
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index cb787b4e3e8bb8da89b2a1eb9b040e78edeccf09..6abbab4d907c9bbf0d7f622428ab096302699915 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2683,4 +2683,10 @@ public final class CraftServer implements Server {
     // Paper end
 	
 	@Override public java.time.Duration getLastTickTime() { return net.minecraft.server.MinecraftServer.lastTickTime; } // Sugarcane
+	
+	@Override
+    public boolean isLagging() {
+        return getServer().lagging;
+    }
+     // Purpur end
 }
diff --git a/src/main/java/xyz/arthurb/mirai/MiraiConfig.java b/src/main/java/xyz/arthurb/mirai/MiraiConfig.java
index b444f912ee87a3b4b612f2ce425960d01b2e013c..0e86297d88adff0fcbd5b0abf8e28abf493483c1 100644
--- a/src/main/java/xyz/arthurb/mirai/MiraiConfig.java
+++ b/src/main/java/xyz/arthurb/mirai/MiraiConfig.java
@@ -209,5 +209,11 @@ public class MiraiConfig {
 	private static void protocolLib() {
 		fixProtocolLib = getBoolean("settings.fix-protocollib", fixProtocolLib);
 	}
+	
+	public static double laggingThreshold = 18.0D;
+	
+	private static void tickLoopSettings() {
+        laggingThreshold = getDouble("settings.lagging-threshold", laggingThreshold);
+    }
 
 }
\ No newline at end of file
