From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etil <81570777+etil2jz@users.noreply.github.com>
Date: Sun, 26 Sep 2021 11:40:35 +0200
Subject: [PATCH] Some modifications to the Snooper patch


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 38b6595ac5d709795b8e846e6113d55481ddd79d..7aeeb858e0b44008b48e26ec4c61f04a203ed288 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -120,8 +120,8 @@ import net.minecraft.util.profiling.metrics.profiling.ServerMetricsSamplersProvi
 import net.minecraft.util.profiling.metrics.storage.MetricsPersister;
 import net.minecraft.util.thread.ReentrantBlockableEventLoop;
 import net.minecraft.world.Difficulty;
-import net.minecraft.world.Snooper;
-import net.minecraft.world.SnooperPopulator;
+//import net.minecraft.world.Snooper;
+//import net.minecraft.world.SnooperPopulator;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.ai.village.VillageSiege;
 import net.minecraft.world.entity.npc.CatSpawner;
@@ -1536,10 +1536,10 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<Runnab
         if (((DedicatedServer) this).getProperties().snooperEnabled && this.tickCount % 6000 == 0) { // Spigot
             this.snooper.prepare();
         }
-		*/
 
         io.papermc.paper.util.CachedLists.reset(); // Paper
         this.profiler.pop();
+		*/
 
         // Paper start - move executeAll() into full server tick timing
         try (co.aikar.timings.Timing ignored = MinecraftTimings.processTasksTimer.startTiming()) {
@@ -1931,12 +1931,10 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<Runnab
         snooper.setFixedData("gui_supported", GraphicsEnvironment.isHeadless() ? "headless" : "supported");
         snooper.setFixedData("dedicated", this.isDedicatedServer());
     }
-	*/
 
-    /* // Patina - completely remove snooper
 	@Override
     public boolean isSnooperEnabled() {
-        return true;
+        return false;
     }
 	*/
 
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 02973f4ac85217364a41596748dc17581ecf6bd0..d0ca92837e95483e94faf4115f5de7a4e6adbfc9 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -512,9 +512,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         snooper.setDynamicData("whitelist_count", (Object) this.getPlayerList().getWhiteListNames().length);
         super.populateSnooper(snooper);
     }
-	*/
 
-    /* // Patina - completely remove snooper
 	@Override
     public boolean isSnooperEnabled() {
         return this.getProperties().snooperEnabled;
