From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Thu, 23 Dec 2021 23:59:12 -0500
Subject: [PATCH] PaperPR Fix entity position desync


diff --git a/src/main/java/net/minecraft/server/level/ServerEntity.java b/src/main/java/net/minecraft/server/level/ServerEntity.java
index 4f332b1c332e7b7ac4ed566d07f36e32b36cf381..bc64a0810f0b90b9baae05d63fa802cf7ddf13be 100644
--- a/src/main/java/net/minecraft/server/level/ServerEntity.java
+++ b/src/main/java/net/minecraft/server/level/ServerEntity.java
@@ -164,7 +164,7 @@ public class ServerEntity {
                 }
                 // CraftBukkit end
 
-                if (this.tickCount > 0 || this.entity instanceof AbstractArrow) {
+                if (this.tickCount >= 0 || this.entity instanceof AbstractArrow) {
                     // Paper start - remove allocation of Vec3D here
                     long k = ClientboundMoveEntityPacket.entityToPacket(vec3d_dx);
                     long l = ClientboundMoveEntityPacket.entityToPacket(vec3d_dy);
