From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etil <81570777+etil2jz@users.noreply.github.com>
Date: Mon, 3 Jan 2022 16:03:27 +0100
Subject: [PATCH] Configurable server metrics


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index 107809776c08a005fa3f0346e42c9d92e753c411..6fe6fc63d712e8c3f71af54cfd253977cccbe33d 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -123,7 +123,7 @@ public class PaperConfig {
             MinecraftServer.getServer().server.getCommandMap().register(entry.getKey(), "Paper", entry.getValue());
         }
 
-        if (!metricsStarted) {
+        if (!metricsStarted && wtf.etil.mirai.MiraiConfig.serverMetrics) { // Mirai - configurable server metrics
             Metrics.PaperMetrics.startMetrics();
             metricsStarted = true;
         }
diff --git a/src/main/java/wtf/etil/mirai/MiraiConfig.java b/src/main/java/wtf/etil/mirai/MiraiConfig.java
index 551cf30c52bb55e7c36cc27f3700c5cd27809c35..1cf3fe75ddc6fee55425ce5b242c370d6b2367d2 100644
--- a/src/main/java/wtf/etil/mirai/MiraiConfig.java
+++ b/src/main/java/wtf/etil/mirai/MiraiConfig.java
@@ -196,6 +196,9 @@ public class MiraiConfig {
 		checkVehicleFlying = getBoolean("checks.vehicle-flight", checkVehicleFlying);
 	}
 
-
+    public static boolean serverMetrics = false;
+    private static void allowServerMetrics() {
+        serverMetrics = getBoolean("enable-server-metrics", serverMetrics);
+    }
 
 }
\ No newline at end of file
