From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: etil2jz <blanchot.arthur@protonmail.ch>
Date: Fri, 4 Mar 2022 19:57:07 +0100
Subject: [PATCH] Skip cloning loot parameters


diff --git a/src/main/java/net/minecraft/advancements/Advancement.java b/src/main/java/net/minecraft/advancements/Advancement.java
index c534055f90b98942b9d7d152fd502cf5387c995b..9e46acf66df7a41eb80d7e41871c3b3caa66f55e 100644
--- a/src/main/java/net/minecraft/advancements/Advancement.java
+++ b/src/main/java/net/minecraft/advancements/Advancement.java
@@ -46,7 +46,7 @@ public class Advancement {
     public Advancement(ResourceLocation id, @Nullable Advancement parent, @Nullable DisplayInfo display, AdvancementRewards rewards, Map<String, Criterion> criteria, String[][] requirements) {
         this.id = id;
         this.display = display;
-        this.criteria = ImmutableMap.copyOf(criteria);
+        this.criteria = java.util.Collections.unmodifiableMap(criteria); // Mirai - Skip cloning loot parameters
         this.parent = parent;
         this.rewards = rewards;
         this.requirements = requirements;
