From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etil <81570777+etil2jz@users.noreply.github.com>
Date: Mon, 3 Jan 2022 16:03:27 +0100
Subject: [PATCH] Configurable server metrics


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index 560e83f1e70cc2712e9c08ffd5e2cab8a6de687b..5c1aee1e41dbf32f01d5278b4942d5bfde1d05e0 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -123,7 +123,7 @@ public class PaperConfig {
             MinecraftServer.getServer().server.getCommandMap().register(entry.getKey(), "Paper", entry.getValue());
         }
 
-        if (!metricsStarted) {
+        if (!metricsStarted && wtf.etil.mirai.MiraiConfig.serverMetrics) { // Mirai - configurable server metrics
             Metrics.PaperMetrics.startMetrics();
             metricsStarted = true;
         }
diff --git a/src/main/java/wtf/etil/mirai/MiraiConfig.java b/src/main/java/wtf/etil/mirai/MiraiConfig.java
index a1d88b8cadf6c193deb69bb577936ebd69e63b2a..21c4a287c8fc550be36fc5b60bdc458d71a9d25d 100644
--- a/src/main/java/wtf/etil/mirai/MiraiConfig.java
+++ b/src/main/java/wtf/etil/mirai/MiraiConfig.java
@@ -230,4 +230,9 @@ public class MiraiConfig {
         entitiesCanUsePortals = getBoolean("entities-can-use-portals", entitiesCanUsePortals);
     }
 
+    public static boolean serverMetrics = true;
+    private static void allowServerMetrics() {
+        serverMetrics = getBoolean("enable-server-metrics", serverMetrics);
+    }
+
 }
\ No newline at end of file
