From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etil <81570777+etil2jz@users.noreply.github.com>
Date: Mon, 3 Jan 2022 16:03:27 +0100
Subject: [PATCH] Configurable server metrics


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index 5ffc636329029ae070e7a79ab9c822c17c0c5fa6..7bbc7ee429716217971fab5ebc0f5f3aa5ae22be 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -123,7 +123,7 @@ public class PaperConfig {
             MinecraftServer.getServer().server.getCommandMap().register(entry.getKey(), "Paper", entry.getValue());
         }
 
-        if (!metricsStarted) {
+        if (!metricsStarted && wtf.etil.mirai.MiraiConfig.serverMetrics) { // Mirai - configurable server metrics
             Metrics.PaperMetrics.startMetrics();
             metricsStarted = true;
         }
diff --git a/src/main/java/wtf/etil/mirai/MiraiConfig.java b/src/main/java/wtf/etil/mirai/MiraiConfig.java
index 9a4b49c6f4a96892fbca94cd6b6203a026c6d66b..370b80c9c35fd314fc51a072d778191e101bcdee 100644
--- a/src/main/java/wtf/etil/mirai/MiraiConfig.java
+++ b/src/main/java/wtf/etil/mirai/MiraiConfig.java
@@ -237,4 +237,9 @@ public class MiraiConfig {
         signAllowColors = getBoolean("sign.allow-colors", signAllowColors);
     }
 
+    public static boolean serverMetrics = true;
+    private static void allowServerMetrics() {
+        serverMetrics = getBoolean("enable-server-metrics", serverMetrics);
+    }
+
 }
\ No newline at end of file
