From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etil <81570777+etil2jz@users.noreply.github.com>
Date: Mon, 3 Jan 2022 16:03:27 +0100
Subject: [PATCH] Configurable server metrics


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index 38bd26b1dff4cb4f71a12b0e84f8133c28024eff..5e58f803d220e13c56f44c846db9af01d94ac974 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -123,7 +123,7 @@ public class PaperConfig {
             MinecraftServer.getServer().server.getCommandMap().register(entry.getKey(), "Paper", entry.getValue());
         }
 
-        if (!metricsStarted) {
+        if (!metricsStarted && wtf.etil.mirai.MiraiConfig.serverMetrics) { // Mirai - configurable server metrics
             Metrics.PaperMetrics.startMetrics();
             metricsStarted = true;
         }
diff --git a/src/main/java/wtf/etil/mirai/MiraiConfig.java b/src/main/java/wtf/etil/mirai/MiraiConfig.java
index 726c0fdb46619e0e0c57b7b5a65be4270acdd0eb..42e13923dd00cfcc95280b17a79a11d1567bcd6a 100644
--- a/src/main/java/wtf/etil/mirai/MiraiConfig.java
+++ b/src/main/java/wtf/etil/mirai/MiraiConfig.java
@@ -237,4 +237,9 @@ public class MiraiConfig {
         signAllowColors = getBoolean("sign.allow-colors", signAllowColors);
     }
 
+    public static boolean serverMetrics = false;
+    private static void allowServerMetrics() {
+        serverMetrics = getBoolean("enable-server-metrics", serverMetrics);
+    }
+
 }
\ No newline at end of file
