From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Thu, 23 Dec 2021 23:59:12 -0500
Subject: [PATCH] PaperPR Fix entity position desync


diff --git a/src/main/java/net/minecraft/server/level/ServerEntity.java b/src/main/java/net/minecraft/server/level/ServerEntity.java
index 756d4a9f7e2be49076af43212a7c1126e28a11fc..ab0ec6d91694abb836a70e7ce2971f63913a53f6 100644
--- a/src/main/java/net/minecraft/server/level/ServerEntity.java
+++ b/src/main/java/net/minecraft/server/level/ServerEntity.java
@@ -164,7 +164,7 @@ public class ServerEntity {
                 }
                 // CraftBukkit end
 
-                if (this.tickCount > 0 || this.entity instanceof AbstractArrow) {
+                if (this.tickCount >= 0 || this.entity instanceof AbstractArrow) {
                     // Paper start - remove allocation of Vec3D here
                     long k = ClientboundMoveEntityPacket.entityToPacket(vec3d_dx);
                     long l = ClientboundMoveEntityPacket.entityToPacket(vec3d_dy);
