From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: etil2jz <blanchot.arthur@protonmail.ch>
Date: Sat, 9 Apr 2022 00:53:41 +0200
Subject: [PATCH] Tweak default values from yml files


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index c57b920d7ca23f3a0e3e10bfe8a4aea7b98c2c6e..e454389be69e082c0e59e8027efd0eb973653cd4 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -251,11 +251,11 @@ public class PaperConfig {
         timings = reallyEnableTimings;
         // Pufferfish end
         boolean verboseTimings = getBoolean("timings.verbose", true);
-        TimingsManager.url = getString("timings.url", "https://timings.aikar.co/");
+        TimingsManager.url = getString("timings.url", "https://timin.gs/");
         // Purpur start
         if (org.purpurmc.purpur.PurpurConfig.version < 23 && TimingsManager.url.contains("timings.pl3x.net")) {
-            set("timings.url", "https://timings.aikar.co/");
-            TimingsManager.url = "https://timings.aikar.co/";
+            set("timings.url", "https://timin.gs/");
+            TimingsManager.url = "https://timin.gs/";
         }
         // Purpur end
         if (!TimingsManager.url.endsWith("/")) {
@@ -289,9 +289,9 @@ public class PaperConfig {
         loadPermsBeforePlugins = getBoolean("settings.load-permissions-yml-before-plugins", true);
     }
 
-    public static int regionFileCacheSize = 256;
+    public static int regionFileCacheSize = 512;
     private static void regionFileCacheSize() {
-        regionFileCacheSize = Math.max(getInt("settings.region-file-cache-size", 256), 4);
+        regionFileCacheSize = Math.max(getInt("settings.region-file-cache-size", 512), 4);
     }
 
     public static boolean enablePlayerCollisions = true;
@@ -368,9 +368,9 @@ public class PaperConfig {
         }
     }
 
-    public static boolean useAlternativeLuckFormula = false;
+    public static boolean useAlternativeLuckFormula = true;
     private static void useAlternativeLuckFormula() {
-        useAlternativeLuckFormula = getBoolean("settings.use-alternative-luck-formula", false);
+        useAlternativeLuckFormula = getBoolean("settings.use-alternative-luck-formula", true);
         if (useAlternativeLuckFormula) {
             Bukkit.getLogger().log(Level.INFO, "Using Aikar's Alternative Luck Formula to apply Luck attribute to all loot pool calculations. See https://luckformula.emc.gs");
         }
@@ -419,7 +419,7 @@ public class PaperConfig {
         }
     }
 
-    public static int maxBookPageSize = 2560;
+    public static int maxBookPageSize = 400;
     public static double maxBookTotalSizeMultiplier = 0.98D;
     private static void maxBookSize() {
         maxBookPageSize = Math.min(8192, getInt("settings.book-size.page-max", maxBookPageSize));
diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index fba36a0fd7036ba516e14c8d3521b470b0df17d1..026c6ac268a1f8678be764a2f6f699026c87b90a 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -153,7 +153,7 @@ public class PaperWorldConfig {
         wanderingTraderSpawnChanceMax = getInt("wandering-trader.spawn-chance-max", wanderingTraderSpawnChanceMax);
     }
 
-    public boolean fixClimbingBypassingCrammingRule = false;
+    public boolean fixClimbingBypassingCrammingRule = true;
     private void fixClimbingBypassingCrammingRule() {
         fixClimbingBypassingCrammingRule = getBoolean("fix-climbing-bypassing-cramming-rule", fixClimbingBypassingCrammingRule);
     }
@@ -191,12 +191,24 @@ public class PaperWorldConfig {
 
     public Map<net.minecraft.world.entity.EntityType<?>, Integer> entityPerChunkSaveLimits = new HashMap<>();
     private void entityPerChunkSaveLimits() {
-        getInt("entity-per-chunk-save-limit.experience_orb", -1);
-        getInt("entity-per-chunk-save-limit.snowball", -1);
-        getInt("entity-per-chunk-save-limit.ender_pearl", -1);
-        getInt("entity-per-chunk-save-limit.arrow", -1);
-        getInt("entity-per-chunk-save-limit.fireball", -1);
-        getInt("entity-per-chunk-save-limit.small_fireball", -1);
+        getInt("entity-per-chunk-save-limit.experience_orb", 16);
+        getInt("entity-per-chunk-save-limit.snowball", 8);
+        getInt("entity-per-chunk-save-limit.ender_pearl", 8);
+        getInt("entity-per-chunk-save-limit.arrow", 16);
+        getInt("entity-per-chunk-save-limit.fireball", 8);
+        getInt("entity-per-chunk-save-limit.small_fireball", 8);
+        getInt("entity-per-chunk-save-limit.dragon_fireball", 3);
+        getInt("entity-per-chunk-save-limit.egg", 8);
+        getInt("entity-per-chunk-save-limit.eye_of_ender", 8);
+        getInt("entity-per-chunk-save-limit.firework_rocket", 8);
+        getInt("entity-per-chunk-save-limit.potion", 8);
+        getInt("entity-per-chunk-save-limit.llama_spit", 3);
+        getInt("entity-per-chunk-save-limit.shulker_bullet", 8);
+        getInt("entity-per-chunk-save-limit.spectral_arrow", 16);
+        getInt("entity-per-chunk-save-limit.experience_bottle", 3);
+        getInt("entity-per-chunk-save-limit.trident", 16);
+        getInt("entity-per-chunk-save-limit.wither_skull", 4);
+        getInt("entity-per-chunk-save-limit.area_effect_cloud", 8);
 
         addEntityPerChunkSaveLimitsFromSection(config.getConfigurationSection("world-settings.default.entity-per-chunk-save-limit"), entityPerChunkSaveLimits);
         addEntityPerChunkSaveLimitsFromSection(config.getConfigurationSection("world-settings." + worldName + ".entity-per-chunk-save-limit"), entityPerChunkSaveLimits);
@@ -227,9 +239,9 @@ public class PaperWorldConfig {
         log( "Keep Spawn Loaded Range: " + (keepLoadedRange/16));
     }
 
-    public int autoSavePeriod = -1;
+    public int autoSavePeriod = 6000;
     private void autoSavePeriod() {
-        autoSavePeriod = getInt("auto-save-interval", -1);
+        autoSavePeriod = getInt("auto-save-interval", 6000);
         if (autoSavePeriod > 0) {
             log("Auto Save Interval: " +autoSavePeriod + " (" + (autoSavePeriod / 20) + "s)");
         } else if (autoSavePeriod < 0) {
@@ -237,9 +249,9 @@ public class PaperWorldConfig {
         }
     }
 
-    public int maxAutoSaveChunksPerTick = 24;
+    public int maxAutoSaveChunksPerTick = 8;
     private void maxAutoSaveChunksPerTick() {
-        maxAutoSaveChunksPerTick = getInt("max-auto-save-chunks-per-tick", 24);
+        maxAutoSaveChunksPerTick = getInt("max-auto-save-chunks-per-tick", 8);
     }
 
     private boolean getBoolean(String path, boolean def) {
@@ -321,7 +333,7 @@ public class PaperWorldConfig {
 
     public boolean nerfedMobsShouldJump;
     private void nerfedMobsShouldJump() {
-        nerfedMobsShouldJump = getBoolean("spawner-nerfed-mobs-should-jump", false);
+        nerfedMobsShouldJump = getBoolean("spawner-nerfed-mobs-should-jump", true);
     }
 
     public final Reference2IntMap<MobCategory> softDespawnDistances = new Reference2IntOpenHashMap<>(MobCategory.values().length);
@@ -396,7 +408,7 @@ public class PaperWorldConfig {
 
     public boolean optimizeExplosions;
     private void optimizeExplosions() {
-        optimizeExplosions = getBoolean("optimize-explosions", false);
+        optimizeExplosions = getBoolean("optimize-explosions", true);
         log("Optimize explosions: " + optimizeExplosions);
     }
 
@@ -417,7 +429,7 @@ public class PaperWorldConfig {
 
     public int mobSpawnerTickRate;
     private void mobSpawnerTickRate() {
-        mobSpawnerTickRate = getInt("mob-spawner-tick-rate", 1);
+        mobSpawnerTickRate = getInt("mob-spawner-tick-rate", 2);
     }
 
     public Integer waterAnimalMaxSpawnHeight;
@@ -440,12 +452,12 @@ public class PaperWorldConfig {
 
     public int containerUpdateTickRate;
     private void containerUpdateTickRate() {
-        containerUpdateTickRate = getInt("container-update-tick-rate", 1);
+        containerUpdateTickRate = getInt("container-update-tick-rate", 3);
     }
 
     public boolean disableChestCatDetection;
     private void disableChestCatDetection() {
-        disableChestCatDetection = getBoolean("game-mechanics.disable-chest-cat-detection", false);
+        disableChestCatDetection = getBoolean("game-mechanics.disable-chest-cat-detection", true);
     }
 
     public boolean disablePlayerCrits;
@@ -487,14 +499,14 @@ public class PaperWorldConfig {
         nonPlayerEntitiesOnScoreboards = getBoolean("allow-non-player-entities-on-scoreboards", false);
     }
 
-    public int nonPlayerArrowDespawnRate = -1;
-    public int creativeArrowDespawnRate = -1;
+    public int nonPlayerArrowDespawnRate = 20;
+    public int creativeArrowDespawnRate = 20;
     private void nonPlayerArrowDespawnRate() {
-        nonPlayerArrowDespawnRate = getInt("non-player-arrow-despawn-rate", -1);
+        nonPlayerArrowDespawnRate = getInt("non-player-arrow-despawn-rate", 20);
         if (nonPlayerArrowDespawnRate == -1) {
             nonPlayerArrowDespawnRate = spigotConfig.arrowDespawnRate;
         }
-        creativeArrowDespawnRate = getInt("creative-arrow-despawn-rate", -1);
+        creativeArrowDespawnRate = getInt("creative-arrow-despawn-rate", 20);
         if (creativeArrowDespawnRate == -1) {
             creativeArrowDespawnRate = spigotConfig.arrowDespawnRate;
         }
@@ -520,7 +532,7 @@ public class PaperWorldConfig {
         fixedInhabitedTime = getInt("fixed-chunk-inhabited-time", -1);
     }
 
-    public int grassUpdateRate = 1;
+    public int grassUpdateRate = 4;
     private void grassUpdateRate() {
         grassUpdateRate = Math.max(0, getInt("grass-spread-tick-rate", grassUpdateRate));
         log("Grass Spread Tick Rate: " + grassUpdateRate);
@@ -591,7 +603,7 @@ public class PaperWorldConfig {
         }
     }
 
-    public int maxCollisionsPerEntity = 8;
+    public int maxCollisionsPerEntity = 2;
     private void maxEntityCollision() {
         maxCollisionsPerEntity = getInt( "max-entity-collisions", this.spigotConfig.getInt("max-entity-collisions", this.maxCollisionsPerEntity, false) );
         log( "Max Entity Collisions: " + maxCollisionsPerEntity );
@@ -630,17 +642,17 @@ public class PaperWorldConfig {
         scanForLegacyEnderDragon = getBoolean("game-mechanics.scan-for-legacy-ender-dragon", true);
     }
 
-    public boolean ironGolemsCanSpawnInAir = false;
+    public boolean ironGolemsCanSpawnInAir = true;
     private void ironGolemsCanSpawnInAir() {
         ironGolemsCanSpawnInAir = getBoolean("iron-golems-can-spawn-in-air", ironGolemsCanSpawnInAir);
     }
 
-    public boolean armorStandEntityLookups = true;
+    public boolean armorStandEntityLookups = false;
     private void armorStandEntityLookups() {
-        armorStandEntityLookups = getBoolean("armor-stands-do-collision-entity-lookups", true);
+        armorStandEntityLookups = getBoolean("armor-stands-do-collision-entity-lookups", false);
     }
 
-    public boolean armorStandTick = true;
+    public boolean armorStandTick = false;
     private void armorStandTick() {
         this.armorStandTick = this.getBoolean("armor-stands-tick", this.armorStandTick);
         log("ArmorStand ticking is " + (this.armorStandTick ? "enabled" : "disabled") + " by default");
@@ -652,9 +664,9 @@ public class PaperWorldConfig {
         log("Water over lava flow speed: " + waterOverLavaFlowSpeed);
     }
 
-    public boolean preventMovingIntoUnloadedChunks = false;
+    public boolean preventMovingIntoUnloadedChunks = true;
     private void preventMovingIntoUnloadedChunks() {
-        preventMovingIntoUnloadedChunks = getBoolean("prevent-moving-into-unloaded-chunks", false);
+        preventMovingIntoUnloadedChunks = getBoolean("prevent-moving-into-unloaded-chunks", true);
     }
 
     public enum DuplicateUUIDMode {
@@ -737,8 +749,8 @@ public class PaperWorldConfig {
     }
 
     public boolean cooldownHopperWhenFull = true;
-    public boolean disableHopperMoveEvents = false;
-    public boolean hoppersIgnoreOccludingBlocks = false;
+    public boolean disableHopperMoveEvents = true;
+    public boolean hoppersIgnoreOccludingBlocks = true;
     private void hopperOptimizations() {
         cooldownHopperWhenFull = getBoolean("hopper.cooldown-when-full", cooldownHopperWhenFull);
         log("Cooldown Hoppers when Full: " + (cooldownHopperWhenFull ? "enabled" : "disabled"));
@@ -911,7 +923,7 @@ public class PaperWorldConfig {
         enderDragonsDeathAlwaysPlacesDragonEgg = getBoolean("ender-dragons-death-always-places-dragon-egg", enderDragonsDeathAlwaysPlacesDragonEgg);
     }
 
-    public boolean updatePathfindingOnBlockUpdate = true;
+    public boolean updatePathfindingOnBlockUpdate = false;
     private void setUpdatePathfindingOnBlockUpdate() {
         updatePathfindingOnBlockUpdate = getBoolean("update-pathfinding-on-block-update", this.updatePathfindingOnBlockUpdate);
     }
@@ -939,7 +951,7 @@ public class PaperWorldConfig {
 
     public long delayChunkUnloadsBy;
     private void delayChunkUnloadsBy() {
-        delayChunkUnloadsBy = PaperConfig.getSeconds(getString("delay-chunk-unloads-by", "10s"));
+        delayChunkUnloadsBy = PaperConfig.getSeconds(getString("delay-chunk-unloads-by", "5s"));
         if (delayChunkUnloadsBy > 0) {
             log("Delaying chunk unloads by " + delayChunkUnloadsBy + " seconds");
             delayChunkUnloadsBy *= 20;
@@ -989,8 +1001,13 @@ public class PaperWorldConfig {
     private com.google.common.collect.Table<String, String, Integer> sensorTickRates;
     private com.google.common.collect.Table<String, String, Integer> behaviorTickRates;
     private void tickRates() {
-        config.addDefault("world-settings.default.tick-rates.sensor.villager.secondarypoisensor", 40);
-        config.addDefault("world-settings.default.tick-rates.behavior.villager.validatenearbypoi", -1); // Example
+        config.addDefault("world-settings.default.tick-rates.sensor.villager.secondarypoisensor", 80);
+        config.addDefault("world-settings.default.tick-rates.sensor.villager.nearestbedsensor", 80);
+        config.addDefault("world-settings.default.tick-rates.sensor.villager.villagerbabiessensor", 40);
+        config.addDefault("world-settings.default.tick-rates.sensor.villager.playersensor", 40);
+        config.addDefault("world-settings.default.tick-rates.sensor.villager.nearestlivingentitysensor", 40);
+        config.addDefault("world-settings.default.tick-rates.behavior.villager.validatenearbypoi", 60); // Example
+        config.addDefault("world-settings.default.tick-rates.behavior.villager.acquirepoi", 120);
         log("Tick rates:");
         sensorTickRates = loadTickRates("sensor");
         behaviorTickRates = loadTickRates("behavior");
diff --git a/src/main/java/gg/pufferfish/pufferfish/PufferfishConfig.java b/src/main/java/gg/pufferfish/pufferfish/PufferfishConfig.java
index a820bfe38cb5765ee74c4fd7084d949477f83692..ac3f6b745a9f8663bde57c77ed0d7330ca8a2a5f 100644
--- a/src/main/java/gg/pufferfish/pufferfish/PufferfishConfig.java
+++ b/src/main/java/gg/pufferfish/pufferfish/PufferfishConfig.java
@@ -213,8 +213,8 @@ public class PufferfishConfig {
     public static int maxProjectileLoadsPerTick;
     public static int maxProjectileLoadsPerProjectile;
     private static void projectileLoading() {
-        maxProjectileLoadsPerTick = getInt("projectile.max-loads-per-tick", 10, "Controls how many chunks are allowed", "to be sync loaded by projectiles in a tick.");
-        maxProjectileLoadsPerProjectile = getInt("projectile.max-loads-per-projectile", 10, "Controls how many chunks a projectile", "can load in its lifetime before it gets", "automatically removed.");
+        maxProjectileLoadsPerTick = getInt("projectile.max-loads-per-tick", 8, "Controls how many chunks are allowed", "to be sync loaded by projectiles in a tick.");
+        maxProjectileLoadsPerProjectile = getInt("projectile.max-loads-per-projectile", 8, "Controls how many chunks a projectile", "can load in its lifetime before it gets", "automatically removed.");
 
         setComment("projectile", "Optimizes projectile settings");
     }
diff --git a/src/main/java/org/spigotmc/SpigotConfig.java b/src/main/java/org/spigotmc/SpigotConfig.java
index d509187ca63963fdd7f1a44d89d2aa1a1b1ce3bd..8a2244d89015066cd10b090b085793241feedc13 100644
--- a/src/main/java/org/spigotmc/SpigotConfig.java
+++ b/src/main/java/org/spigotmc/SpigotConfig.java
@@ -335,7 +335,7 @@ public class SpigotConfig
     public static boolean saveUserCacheOnStopOnly;
     private static void saveUserCacheOnStopOnly()
     {
-        SpigotConfig.saveUserCacheOnStopOnly = SpigotConfig.getBoolean( "settings.save-user-cache-on-stop-only", false );
+        SpigotConfig.saveUserCacheOnStopOnly = SpigotConfig.getBoolean( "settings.save-user-cache-on-stop-only", true );
     }
 
     public static double movedWronglyThreshold;
diff --git a/src/main/java/org/spigotmc/SpigotWorldConfig.java b/src/main/java/org/spigotmc/SpigotWorldConfig.java
index c405457a775df902540602dfb4cdcd6872d35e2a..1b877e09da06b55e90002399c6a976b1db80505c 100644
--- a/src/main/java/org/spigotmc/SpigotWorldConfig.java
+++ b/src/main/java/org/spigotmc/SpigotWorldConfig.java
@@ -150,14 +150,14 @@ public class SpigotWorldConfig
     public double itemMerge;
     private void itemMerge()
     {
-        this.itemMerge = this.getDouble("merge-radius.item", 2.5 );
+        this.itemMerge = this.getDouble("merge-radius.item", 3.5 );
         this.log( "Item Merge Radius: " + this.itemMerge );
     }
 
     public double expMerge;
     private void expMerge()
     {
-        this.expMerge = this.getDouble("merge-radius.exp", 3.0 );
+        this.expMerge = this.getDouble("merge-radius.exp", 4.0 );
         this.log( "Experience Merge Radius: " + this.expMerge );
     }
 
@@ -207,31 +207,31 @@ public class SpigotWorldConfig
         this.log( "Item Despawn Rate: " + this.itemDespawnRate );
     }
 
-    public int animalActivationRange = 32;
-    public int monsterActivationRange = 32;
+    public int animalActivationRange = 16;
+    public int monsterActivationRange = 24;
     public int raiderActivationRange = 48;
-    public int miscActivationRange = 16;
+    public int miscActivationRange = 8;
     // Paper start
-    public int flyingMonsterActivationRange = 32;
-    public int waterActivationRange = 16;
-    public int villagerActivationRange = 32;
-    public int wakeUpInactiveAnimals = 4;
+    public int flyingMonsterActivationRange = 48;
+    public int waterActivationRange = 8;
+    public int villagerActivationRange = 16;
+    public int wakeUpInactiveAnimals = 2;
     public int wakeUpInactiveAnimalsEvery = 60*20;
-    public int wakeUpInactiveAnimalsFor = 5*20;
-    public int wakeUpInactiveMonsters = 8;
+    public int wakeUpInactiveAnimalsFor = 2*20;
+    public int wakeUpInactiveMonsters = 4;
     public int wakeUpInactiveMonstersEvery = 20*20;
-    public int wakeUpInactiveMonstersFor = 5*20;
-    public int wakeUpInactiveVillagers = 4;
+    public int wakeUpInactiveMonstersFor = 3*20;
+    public int wakeUpInactiveVillagers = 1;
     public int wakeUpInactiveVillagersEvery = 30*20;
-    public int wakeUpInactiveVillagersFor = 5*20;
-    public int wakeUpInactiveFlying = 8;
+    public int wakeUpInactiveVillagersFor = 20;
+    public int wakeUpInactiveFlying = 1;
     public int wakeUpInactiveFlyingEvery = 10*20;
-    public int wakeUpInactiveFlyingFor = 5*20;
+    public int wakeUpInactiveFlyingFor = 3*20;
     public int villagersWorkImmunityAfter = 5*20;
     public int villagersWorkImmunityFor = 20;
     public boolean villagersActiveForPanic = true;
     // Paper end
-    public boolean tickInactiveVillagers = true;
+    public boolean tickInactiveVillagers = false;
     public boolean ignoreSpectatorActivation = false;
     private void activationRange()
     {
@@ -305,7 +305,7 @@ public class SpigotWorldConfig
     public int tridentDespawnRate;
     private void arrowDespawnRate()
     {
-        this.arrowDespawnRate = this.getInt( "arrow-despawn-rate", 1200 );
+        this.arrowDespawnRate = this.getInt( "arrow-despawn-rate", 300 );
         this.tridentDespawnRate = this.getInt( "trident-despawn-rate", this.arrowDespawnRate );
         this.log( "Arrow Despawn Rate: " + this.arrowDespawnRate + " Trident Respawn Rate:" + this.tridentDespawnRate );
     }
diff --git a/src/main/resources/configurations/bukkit.yml b/src/main/resources/configurations/bukkit.yml
index 97239f7aadd761f6f709a6d6c913e577651a87df..3ed0852d0d891c3f0a9270e0c14473b3dab893f2 100644
--- a/src/main/resources/configurations/bukkit.yml
+++ b/src/main/resources/configurations/bukkit.yml
@@ -34,11 +34,11 @@ chunk-gc:
     period-in-ticks: 600
 ticks-per:
     animal-spawns: 400
-    monster-spawns: 1
-    water-spawns: 1
-    water-ambient-spawns: 1
-    water-underground-creature-spawns: 1
-    axolotl-spawns: 1
-    ambient-spawns: 1
+    monster-spawns: 10
+    water-spawns: 400
+    water-ambient-spawns: 400
+    water-underground-creature-spawns: 400
+    axolotl-spawns: 400
+    ambient-spawns: 400
     autosave: 6000
 aliases: now-in-commands.yml
