From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: etil2jz <blanchot.arthur@protonmail.ch>
Date: Sat, 26 Mar 2022 18:48:55 +0100
Subject: [PATCH] Fix SimpleYAML build


diff --git a/build.gradle.kts b/build.gradle.kts
index 97a13bda8f5bc1464cc9fbf7b17f09f5757d33f2..5e75485694263c16dc09411ebf067479aa94d49d 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -59,7 +59,7 @@ dependencies {
 
     // Pufferfish start
     implementation("org.yaml:snakeyaml:1.28")
-    implementation ("me.carleslc.Simple-YAML:Simple-Yaml:1.7.3") {
+    implementation ("me.carleslc.Simple-YAML:Simple-Yaml:1.8") { // Mirai - fix Purpur build
         exclude(group="org.yaml", module="snakeyaml")
     }
     // Pufferfish end
