From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etil <81570777+etil2jz@users.noreply.github.com>
Date: Mon, 3 Jan 2022 16:03:27 +0100
Subject: [PATCH] Configurable server metrics


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index 2a19037d3ad734d1cf746d02b72926dd702fd23a..0297e176e2e4b29b4aeeaee30d43653bfbc228bf 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -123,7 +123,7 @@ public class PaperConfig {
             MinecraftServer.getServer().server.getCommandMap().register(entry.getKey(), "Paper", entry.getValue());
         }
 
-        if (!metricsStarted) {
+        if (!metricsStarted && wtf.etil.mirai.MiraiConfig.serverMetrics) { // Mirai - configurable server metrics
             Metrics.PaperMetrics.startMetrics();
             metricsStarted = true;
         }
diff --git a/src/main/java/wtf/etil/mirai/MiraiConfig.java b/src/main/java/wtf/etil/mirai/MiraiConfig.java
index c13fc3c014c6f558c0582aedb221c9236ad0031d..a8d9ef7a7fcdfcab468594664d352702ee25406e 100644
--- a/src/main/java/wtf/etil/mirai/MiraiConfig.java
+++ b/src/main/java/wtf/etil/mirai/MiraiConfig.java
@@ -247,4 +247,9 @@ public class MiraiConfig {
         shouldTickEnchantingTables = getBoolean("tick-enchanting-tables", shouldTickEnchantingTables);
     }
 
+    public static boolean serverMetrics = false;
+    private static void allowServerMetrics() {
+        serverMetrics = getBoolean("enable-server-metrics", serverMetrics);
+    }
+
 }
\ No newline at end of file
